man/sysupdate.d.xml

```diff
@@ -0,0 +1,872 @@
+    disk partition, the latter is a file that is downloaded and placed in a regular file in the boot file
+    system (e.g. EFI system partition). Hence, during an update of a hypothetical operating system "foobarOS"
+    to a hypothetical version 47 the following operations should take place:</para>
+
+    <orderedlist>
+      <listitem><para>A file <literal>https://download.example.com/foobarOS_47.root.xz</literal> should be
+      downloaded, decompressed and written to a previously empty partition with GPT partition type UUID
+      4f68bce3-e8cd-4db1-96e7-fbcaf984b709 for x86-64, as per <ulink
+      url="https://systemd.io/DISCOVERABLE_PARTITIONS">Discoverable Partitions
+      Specification</ulink>.</para></listitem>
+
+      <listitem><para>Similar, a file <literal>https://download.example.com/foobarOS_47.verity.xz</literal>
+      should be downloaded, decompressed and written to a previously empty partition with GPT partition type
+      UUID of 2c7357ed-ebd2-46d9-aec1-23d437ec2bf5 (i.e the partition type for Verity integrity information
+      for x86-64 root file systems).</para></listitem>
+
+      <listitem><para>Finally, a file <literal>https://download.example.com/foobarOS_47.efi.xz</literal> (a
+      unified kernel, as per <ulink url="https://systemd.io/BOOT_LOADER_SPECIFICATION">Boot Loader
+      Specification</ulink> Type #2) should be downloaded, decompressed and written to the ESP file system,
+      i.e. to <filename>EFI/Linux/foobarOS_47.efi</filename> in the ESP.</para></listitem>
+    </orderedlist>
+
+    <para>The version-independent generalization of this would be (using the special marker
+    <literal>$v</literal> as wildcard for the version identifier):</para>
+
+    <orderedlist>
+      <listitem><para>A transfer of a file <literal>https://download.example.com/foobarOS_$v.root.xz</literal>
+      → a local, previously empty GPT partition of type 4f68bce3-e8cd-4db1-96e7-fbcaf984b709, with a label to
+      be set to <literal>foobarOS_$v</literal>.</para></listitem>
+
+      <listitem><para>A transfer of a file <literal>https://download.example.com/foobarOS_$v.verity.xz</literal>
+      → a local, previously empty GPT partition of type 2c7357ed-ebd2-46d9-aec1-23d437ec2bf5, with a label to be
+      set to <literal>foobarOS_$v_verity</literal>.</para></listitem>
+
+      <listitem><para>A transfer of a file <literal>https://download.example.com/foobarOS_$v.efi.xz</literal>
+      → a local file <filename>/efi/EFI/Linux/foobarOS_$v.efi</filename>.</para></listitem>
+    </orderedlist>
+
+    <para>An update can only complete if the relevant URLs provide their resources for the same version,
+    i.e. for the same value of <literal>$v</literal>.</para>
+
+    <para>The above may be translated into three <filename>*.conf</filename> files in
+    <filename>sysupdate.d/</filename>, one for each resource to transfer. The <filename>*.conf</filename>
```

Comment:
Since those three are tied together, shouldn't they use the component approach?

Reply:
the idea idea is that multiple transfers are bound together in a component. Each transfer is ultimately a separate HTTP request. This if you do an OS update by installing a root fs + verity partition for it + EFI unified kernel, then these three transfers together make up one component.

You may have as many components as you want, each having any name you choose – but one component is special, and is the default one, the one that has no name, and refers to the OS itself.

So the idea is that when figuring out what resources and what components map to to you have to figure out if things should be strictly updated together or if they can be updated separately. Let's say you have an OS image, that consists of rootfs+verity+efibinary, that's one component since you wouldn't download lets say the verity part of it, without the rootfs part, and similar with the others. OTOH you might have some sysext also installed that adds artwork to the OS. It has a separate release cycle, so that could be a separate component. And then you could have an nspawn container image somewhere with its adjacent nspawn file. So with that you would have:

1. Default component – the OS, consisting of:
    1. Transfer defined in /usr/lib/sysupdate.d/50-verity.conf (makes sure verity partition is downloaded)
    2. Transfer defined in /usr/lib/sysupdate.d/60-root.conf (makes sure root fs partition is downloaded)
    3. Transfer defined in /usr/lib/sysupdate.d/70-kernel.conf (makes sure EFI kernel image is downloaded)
2. Component "artwork", consisting of:
    1. Transfer defined in /usr/lib/sysupdate.artwork.d/50-artwork.conf
3. Component "mycontainer", consisting of:
    1. Transfer defined in /usr/lib/sysupdate.mycontainer.d/30-image.conf (takes care of the container .raw download)
    2. Transfer defined in /usr/lib/sysupdate.mycontainer.d/40-nspawn.conf (takes care of the matching .nspawn file)

Now, if you don't specify any component on the "systemd-sysupdate" cmdline this would update the default component, i.e. the OS itself. If you specify `-c artwork`, we'd instead update the second component, and if you specify `-c mycontainer` the third.

As the name suggests "sysupdate" really is about updating the system, i.e. the OS, and hence I tried to make updating the "main component" easy and obvious, and to somewhat hide away the fact you can actually have more than one component. i.e it's a feature for those who already mastered the first level of complexity, and want to update other things than just the OS itself.

i.e. i think introducing systems with two many new concepts is problematic, because it's hard to get into. This I figured it would be easier to just pretend at first this was a tool you can update the OS with, full stop. And then if you read the man page closely you'll find that you can actually define multiple sets of transfers and we call them components and each set is independent of each other set.

TLDR: If you want to update two things independently of each other, stick them in separate transfers in separate components. If you want to update them strictly in sync so that only if all of them are fully acquired they shall be activated, then stick them as multiple transfer into the same component. You can freely choose what applies best to your usecase.

Reply:
I'd really like the "components" stuff to be secondary concepts. I'd expect 90% of the cases would just do full system updates, and that's it. If we'd make the components concept mandatory it would complicate things quite a bit. 

Right now you can totally ignore everything about components and get things done. I#d rather not lose that.